pkg/util/audit/audit.go

```diff
@@ -0,0 +1,158 @@
+package audit
+
+import (
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/ARO-RP/pkg/util/audit/schema"
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	auditMetadataCategory    = "auditCategory"
+	auditMetadataCreatedTime = "auditCreatedTime"
+	auditMetadataFullPayload = "auditFullPayload"
+	auditMetadataLogKind     = "logKind"
+	auditMetadataOperation   = "auditOperation"
+	auditMetadataResult      = "auditResult"
+	auditMetadataSource      = "auditSource"
+	auditSourceRP            = "aro-rp"
+
+	ifxAuditCloudVer = 1.0
+	ifxAuditLogKind  = "ifxaudit"
+	ifxAuditName     = "#Ifx.AuditSchema"
+	ifxAuditVersion  = 2.1
+
+	// ifxAuditFlags is a collection of values bit-packed into a 64-bit integer.
+	// These properties describe how the event should be processed by the pipelinei
+	// in an implementation-independent way.
+	ifxAuditFlags = 257
+
+	// ifxAuditPopSample is the effective percentage sample rate for this event at the
+	// time it was generated by a client (either the default value specified in
+	// the event definition, system default or a dynamic override from a cloud
+	// settings service).
+	// not used by Azure
+	ifxAuditPopSample = 0.0
+)
+
+var (
+	// epoch is an unique identifier associated with the current session of the
+	// telemetry library running on the platform. It must be stable during a
+	// session, and has no implied ordering across sessions.
+	epoch string
+
+	// sequence field used to track absolute order of uploaded events, per session.
+	// it is reset when the ARO component is restarted.
+	seqNum      uint64
+	seqNumMutex *sync.Mutex
+)
+
+func init() {
+	epoch = uuid.NewV4().String()
+	seqNum = 1
+	seqNumMutex = new(sync.Mutex)
```

Comment:
1. the zero value of a sync.Mutex is a valid mutex, so you can just do `var seqNumMutex syncMutex` and have done with it
2. `&foo{}` is preferred over `new(foo)` in this codebase.  Standardising makes it easier to grep for allocation of objects of a given type.